runtime/service.go

```diff
@@ -769,6 +785,142 @@ func (s *service) GetVMMetadata(requestCtx context.Context, request *proto.GetVM
+	return balloon, nil
+}
+
+func (s *service) BuildBalloonDeviceOpt(balloon models.Balloon) ([]firecracker.Opt, error) {
+	handler := firecracker.NewCreateBalloonHandler(*balloon.AmountMb, *balloon.DeflateOnOom, balloon.StatsPollingIntervals)
+	opt := []firecracker.Opt{
+		func(m *firecracker.Machine) {
+			m.Handlers.FcInit = m.Handlers.FcInit.AppendAfter(firecracker.CreateMachineHandlerName, handler)
+		},
+	}
+	return opt, nil
+}
+
+// GetBalloonConfig will get configuration for an existing balloon device, before or after machine startup
+func (s *service) GetBalloonConfig(requestCtx context.Context, req *proto.GetBalloonConfigRequest) (*proto.GetBalloonConfigResponse, error) {
+	defer logPanicAndDie(s.logger)
+
+	err := s.waitVMReady()
+	if err != nil {
+		s.logger.WithError(err).Error()
+		return nil, err
+	}
+
+	s.logger.Info("Getting configuration for the balloon device")
+	balloon, err := s.machine.GetBalloonConfig(requestCtx)
+	if err != nil {
+		return nil, errors.Errorf("Fail to get balloon configuration. Please check if you have successfully created a balloon device")
+	}
+	balloonConfig := &proto.FirecrackerBalloonDevice{
+		AmountMb:              *balloon.AmountMb,
+		DeflateOnOom:          *balloon.DeflateOnOom,
+		StatsPollingIntervals: balloon.StatsPollingIntervals,
+	}
+	return &proto.GetBalloonConfigResponse{BalloonConfig: balloonConfig}, err
+}
+
+// UpdateBalloon will update an existing balloon device, before or after machine startup
+func (s *service) UpdateBalloon(requestCtx context.Context, req *proto.UpdateBalloonRequest) (*empty.Empty, error) {
+	defer logPanicAndDie(s.logger)
+
+	err := s.waitVMReady()
+	if err != nil {
+		s.logger.WithError(err).Error()
+		return nil, err
+	}
+
+	s.logger.Infof("Updating balloon memory size, the new amount memory is %d MB", req.AmountMb)
+	if err := s.machine.UpdateBalloon(requestCtx, req.AmountMb); err != nil {
+		err = errors.Wrap(err, "failed to update memory balloon")
+		s.logger.WithError(err).Error()
+		return nil, err
+	}
+
+	return &empty.Empty{}, nil
+}
+
+// GetBalloonStats will return the latest balloon device statistics, only if enabled pre-boot.
+func (s *service) GetBalloonStats(requestCtx context.Context, req *proto.GetBalloonStatsRequest) (*proto.GetBalloonStatsResponse, error) {
+	defer logPanicAndDie(s.logger)
+
+	err := s.waitVMReady()
+	if err != nil {
+		s.logger.WithError(err).Error()
+		return nil, err
+	}
+
+	s.logger.Info("Getting statistics for the balloon device")
+	balloonStats, err := s.machine.GetBalloonStats(requestCtx)
+	if err != nil {
+		err = errors.Wrap(err, "failed to get balloon statistics")
+		s.logger.WithError(err).Error()
+		return nil, err
+	}
+
+	if balloonStats.ActualMb == nil ||
+		balloonStats.ActualPages == nil ||
+		balloonStats.TargetMb == nil ||
+		balloonStats.TargetPages == nil {
+		s.logger.Errorf("One of BalloonStats properties is nil, please check ActualMb, ActualPages, TargetMb, TargetPages")
+	}
```

Comment:
Could we actually point out which stat is `nil`? 

I think it could be worth it but others can argue against that.

Reply:
I just feel it is a good practice to check before deference a pointer. Open to discuss.

Reply:
Are they nil-able on Firecracker's schema?

I agree it is better to check that before de-referencing, but this code only logs error after that. You must not de-reference them if they are nil.

Reply:
Fixed it.